### PR TITLE
Remove "illegal" UnsafePointer casts from the stdlib.

### PR DESCRIPTION
**Pull request by atrick**
This replaces:
https://github.com/apple/swift/pull/3824

Update for SE-0107: UnsafeRawPointer

This adds a "mutating" initialize to UnsafePointer to make
Immutable -> Mutable conversions explicit.

These are quick fixes to stdlib, overlays, and test cases that are necessary
in order to remove arbitrary UnsafePointer conversions.

Many cases can be expressed better up by reworking the surrounding
code, but we first need a working starting point.

### CODE REVIEW

**Comment by atrick** [comment]
I'm putting this up for another review.
This still needs to be updated for the linux build. I'll fix that.
Also, RaceTest and AnyHashable are failing due to String initializers. I'm not sure if that's related to my changes.

**Comment by gribozavr** [comment]
> Also, RaceTest and AnyHashable are failing due to String initializers. I'm not sure if that's related to my changes.

@CodaFi was working on those.

File: stdlib/public/core/Runtime.swift.gyb [hunk 1]
```diff
@@ -260,14 +253,14 @@ public func _swift_stdlib_atomicFetch${operation}Int(
...
+        object: unsafeBitCast(target, to: UnsafeMutablePointer<UInt32>.self),
+        operand: UInt32(bitPattern: Int32(operand)))))
 #elseif arch(x86_64) || arch(arm64) || arch(powerpc64) || arch(powerpc64le) || arch(s390x)
   return Int(Int64(bitPattern:
-    _swift_stdlib_atomicFetch${operation}UInt64(
-      object: UnsafeMutablePointer(target),
-      operand: UInt64(bitPattern: Int64(operand)))))
+      _swift_stdlib_atomicFetch${operation}UInt64(
+        object: unsafeBitCast(target, to: UnsafeMutablePointer<UInt64>.self),
+        operand: UInt64(bitPattern: Int64(operand)))))
```
Comment (gribozavr) [inline stdlib/public/core/Runtime.swift.gyb]: Extra indentation.

**Comment by gribozavr** [comment]
Changes LGTM, thank you!

File: stdlib/public/core/Runtime.swift.gyb (hunk 1 above)
Comment (atrick) [inline stdlib/public/core/Runtime.swift.gyb]: FYI emacs is intentionally following a convention here. I'm not picky about the convention as long as swift mode can be fixed.

**Comment by atrick** [comment]
@swift-ci Please smoke test.

**Comment by atrick** [comment]
@gribozavr Please merge (so I can move on with another PR before the next commit conflicts).

I don't know what the problem is with this package manager test:
 swift-package-tests :: swift-package-init-lib.md

TEST 'swift-package-tests :: swift-package-init-lib.md'

**Comment by gottesmm** [comment]
@atrick That issue is now taken care of (Thanks @abertelrud!). So in the future, you should no thave this issue.

**Comment by atrick** [comment]
If there are no conflicts (there weren't any an hour ago), then this should be merged. There's no point clogging the build queue any further.

**Pull request merged**